rfcs/014 - Rework IO Parallelization.md

```diff
@@ -0,0 +1,151 @@
+
+```alan
+const googleHomepage = get('https://google.com');
+const yahooHomepage = get('https://yahoo.com');
+```
+
+rightly can be parallelized and reduce the concurrency impact because these two calls really have no hidden state under the hood. But, something like:
+
+```alan
+post('https://mywebsite.com/account/1/orders/new', '{"someOrder": "json"}');
+const allOrders = get('https://mywebsite.com/account/1/orders');
+```
+
+would fail to include the newly added order, because it would start requesting all orders while it is trying to create the new order. There is nothing the compiler can do to determine if there is hidden state in these IO operations, so we **cannot** do anything but linear execution *unless* the developer provides us a hint that it is safe.
+
+The Array methods have a better story on how to determine parallelization (or not). Function purity (whether or not it mutates an outer scope) will eventually automatically cause a `map` to be converted into a `mapLin` call to maintain consistent behavior while "pure" functions that have no external impacts can be executed in parallel.
+
+The absolute simplest solution is to just make all IO operations linear by default and "bolt on" parallelization through arrays and maps, eg:
+
+```
+const homepages = ['https://google.com', 'https://yahoo.com'].map(get);
+```
+
+but this is only a small amount of syntactic sugar on top of idiomatic Javascript for the same:
+
+```js
+const googleHomepage = await (await fetch('https://google.com').text());
+const yahooHomepage = await (await fetch('https://yahoo.com').text());
+```
+
+vs
+
+```js
+const homepages = await Promise.all(['https://google.com', 'https://yahoo.com'].map((url) => (await fetch(url)).json()));
+```
+
+There's some syntactic noise in JS, but not that much, and defining a `get` wrapper function eliminates most of it.
+
+```js
+const googleHomepage = await get('https://google.com');
+const yahooHomepage = await get('https://yahoo.com');
+```
+
+```js
+const homepages = await Promise.all(['https://google.com', 'https://yahoo.com'].map(get));
+```
+
+The third opcode type: `unpred_cpu!` gives us an out; if we had a version of every relevant IO opcode where you could opt in to the current semantics and another version that behaved like `unpred_cpu!` does that is the default, then we get cleaner code in both paths. A few considered suffix names for these: `Par`, like `execPar` based on `reducePar`, but where parallelization makes sense with reduce since there's guaranteed to be many calls, the potential advantages involved with calling `execPar` with only one `exec` usage (but perhaps other IO opcode uses that can batch with it) are unclear. `Async` was considered to potentially confuse people with Rust/C#/JS async/await and people would expect Promises/Futures they need to attach a handler or await on. So while not great, the current suffix under consideration is `Eager`. Eg,
```

Comment:
For the sake of completion/posterity, after we talked I thought of `Disorderly` or `Whenever` as alternatives to `Eager`. However, `Eager` is both shorter and makes much more sense semantically